### ISSUE TITLE
Improve error reporting: if / else branches return different types.

### ISSUE THREAD

**Issue by isaacabraham** [Microsoft/visualfsharp#1105]
## What
The following code results in an error message which could be improved upon: -
```fsharp
let y =
    if test > 10 then "test"
    else 123
```

The error is: -

```error FS0001: This expression was expected to have type
    string    
but here has type
    int```

## Why
Newcomers to expression-based conditionals often do not understand that all branches of a conditional must evaluate to the same type of result. The current error message identifies that two types do not match, but does not make clear the *context* in which this is an error.

## How
A better error might be:

```All branches of an if / else expression must return the same type. The other values in this if / else expression all return a value of type `string`.```

If possible, the values of the other branches should be included in the error message (of course, this will not always be possible).

**Comment by yever** [comment]
Isn't it more a matter of `if ... then ... else ...`?
How about something like `The types of the values produced in the "then" branch (string) and the "else" branch (int) of this conditional expression do not match`.

Paraphrased from [here](https://msdn.microsoft.com/en-us/library/dd233231.aspx#Anchor_1).

**Comment by isaacabraham** [comment]
Could be more than two branches.

**Comment by yever** [comment]
True. How about: `The types of the values produced in all "then" (string) and "else" (int) branches of a conditional expression must match` ?
With more branches, it could be: `The types of the values produced in all "then" (string, int, int) and "else" (int) branches of a conditional expression must match`.

**Comment by yever** [comment]
or: `are expected to match`

**Comment by isaacabraham** [comment]
Hmmm... does it matter whether the branch is an if / elif / else branch? They're all branches of the same expression, they should all match.

**Comment by yever** [comment]
@isaacabraham, I guess it does make sense not to split the "then" and "else" branches (as you mentioned [here too](https://github.com/Microsoft/visualfsharp/issues/1106#issuecomment-214566821)). I still think that mentioning the existence of these branches could be helpful for newcomers. I tried adopting @smoothdeveloper 's way of [listing the types](https://github.com/Microsoft/visualfsharp/issues/1106#issuecomment-214008165) in the following suggestion:
`The values produced in all "then" and "else" branches of a conditional expression are expected to have the same type but here have inconsistent types: string 1 occurrence, int 1 occurrence. Make expression type consistent across all branches`

**Comment by isaacabraham** [comment]
Don't understand. I'm not sure where I mentioned separating else / then branches - if I did I shouldn't have, I don't think it's a good idea. But that's IMHO.

**Comment by yever** [comment]
@isaacabraham, I'll try to be clear: I first suggested to split the branches. You said that the [branches shouldn't be split](https://github.com/Microsoft/visualfsharp/issues/1106#issuecomment-214566821). I [agreed with you](https://github.com/Microsoft/visualfsharp/issues/1105#issuecomment-214844758). We are now on the same page.

The only point I was still making is that just saying "branches" might not make a lot of sense to new F# programmers. That's why I thought it could be clearer to write something like 'all "then" and "else" branches' (which is just a longer and clearer way of saying 'all branches'). The actual list of inconsistent types shouldn't be split. 

This is exactly what I suggested:
`The values produced in all "then" and "else" branches of a conditional expression are expected to have the same type but here have inconsistent types: string 1 occurrence, int 1 occurrence. Make expression type consistent across all branches`

**Comment by smoothdeveloper** [comment]
BTW, the same issue can be obtained with seq/array/list expressions, function/match expressions, I think they should all be addressed in same fashion.

**Comment by forki** [comment]
Will do this together with #1104 in https://github.com/Microsoft/visualfsharp/pull/1142

**Comment by KevinRansom** [comment]
@isaacabraham @forki 

resolved by #1142 

````
> let test = 6
- let y =
-     if test > 10 then "test"
-     else 123;;

      else 123;;
  ---------^^^

stdin(18,10): error FS0001: All branches of an 'if' expression must return the same type. This 
expression was expected to have type 'string' but here has type 'int'.
>
````
If you feel there is still more work needed then pleas submit some additional suggestions and or a follow up PR.
